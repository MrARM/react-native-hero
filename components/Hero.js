'use strict';

import React, { Component } from 'react';
import { Image, StyleSheet, View, Dimensions } from 'react-native';

// a -> b
const createPositionStyle = (zIndex=1) => ({
  top: 0,
  zIndex
});

// a -> a
const updateWidthState = (component, window) => {
  if (!component) return;
  component.setState({
    width: window.width
  });
}
// a, b -> b
export default class Hero extends Component {
  static propTypes = {
    renderOverlay: React.PropTypes.func,
    resizeMode: React.PropTypes.string,
    colorOpacity: React.PropTypes.number,
    minHeight: React.PropTypes.number,
    fullWidth: React.PropTypes.bool
  };

  static defaultProps = {
    fullWidth: true
  }

  constructor(props) {
    super(props);

    this.state = {
      source: this.props.source,
      opacity: 0,
      height: this.props.minHeight || undefined,
      resizeMode: this.props.resizeMode || 'cover'
    };
  };

  componentWillReceiveProps(nextProps) {
    if (nextProps.source !== this.props.source) {
      this.setState({
        source: this.props.source
      });
    }
  };

  componentDidMount() {
    this.setState({
      opacity: 1
    });
  };

  renderColorOverlay() {
    let overlayStyles = [{
      height: this.state.height,
      width: '100%',
      backgroundColor: this.props.colorOverlay,
      opacity: this.props.colorOpacity || .30
    }, createPositionStyle(1), { position: 'absolute'}];

    return (this.props.colorOverlay) ?
      <View style={overlayStyles}></View> :
      null;
  };

  renderHeroOverlay() {
    const self = this;
    const transparentBg = { backgroundColor: 'transparent' };
    const contentStyles = (this.props.colorOverlay) ?
      [transparentBg, createPositionStyle(2)] :
      transparentBg;

    const updateViewHeight = (event) => {
      const overlayHeight = event.nativeEvent.layout.height;
      const overlayWidth = event.nativeEvent.layout.width;
      if (self.props.minHeight) return;
      if (overlayHeight !== self.state.height) {
        self.setState({
          height: overlayHeight
        });
      }

      // Initial width state set
      if ((this.props.fullWidth === true) && (!self.state.width)) {
        updateWidthState(self, Dimensions.get('window'));
        Dimensions.addEventListener('change', (window, screen) => updateWidthState(self, window));
      }
    };

    return (
      <View
        onLayout={(event) => updateViewHeight(event)}
        style={contentStyles}>
          {this.props.renderOverlay()}
      </View>
    );
  };

  render() {
    return (
      <View style={{opacity: this.state.opacity}}>
        <Image
          source={this.state.source}
          resizeMode={this.state.resizeMode}
          style={{height: this.state.height, width: (this.state.width || '100%') }}>
            {this.renderHeroOverlay()}
            {this.renderColorOverlay()}
        </Image>
      </View>
    )
  };
}
