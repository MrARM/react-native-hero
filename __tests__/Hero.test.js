import { Text, View } from 'react-native';
import React from 'react';
import Hero, { _createPositionStyle } from '../components/Hero';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

const imageUrl = 'https://catsarethebest.com/cat1';
const overlayTpl =
      <View>
      <Text>React Native Hero</Text>
      <Text>Testing!</Text>
      </View>;

test('MAIN: Render correctly and overlay contains correct content', () => {
  const tree = renderer.create(
    <Hero
    source={{uri: imageUrl}}
    renderOverlay={() => overlayTpl} />
  ).toJSON();

  const imageParent = tree.children[0];
  // First child should be a Image wrapper
  expect(imageParent.type).toBe('Image');
  expect(imageParent.props.source).toEqual({ uri: imageUrl });
  // Default 100% width of Parent
  expect(imageParent.props.style.width).toBe('100%');
  // No color props set, so only overlay() content should be the only child
  expect(imageParent.children.length).toBe(1);

  const overlay = imageParent.children[0].children[0];
  expect(overlay.type).toBe('View');
  // Text elements of the view
  expect(overlay.children.length).toBe(2);
});

test('MAIN: Render with a static image', () => {
  const tree = renderer.create(
    <Hero
      source={require('../__assets__/lion-sample.jpg')}
      renderOverlay={() => overlayTpl} />
  ).toJSON();

  const imageParent = tree.children[0];
  expect(imageParent.type).toBe('Image');
  expect(imageParent.props.source).toBe(1);
});

test('PROPS: Props.minHeight should override the height', () => {
  const tree = renderer.create(
    <Hero
      source={{uri: imageUrl}}
      renderOverlay={() => overlayTpl}
      minHeight={1000} />
    ).toJSON();

  const imageParent = tree.children[0];
  expect(imageParent.props.style.height).toBe(1000)
});

test('PROPS: Props.fullWidth should default to state updates', () => {
  const tree = renderer.create(
    <Hero
      source={{uri: imageUrl}}
      renderOverlay={() => overlayTpl}
    fullWidth={false} />
    ).toJSON();

  const imageParent = tree.children[0];
  expect(imageParent.props.style.fullWidth).toBeUndefined;
});

test('PROPS: Color overlay', () => {
  const tree = renderer.create(
    <Hero
      source={{uri: imageUrl}}
      renderOverlay={() => overlayTpl}
      colorOverlay={'red'}
      colorOpacity={.40}/>
    ).toJSON();

  const imageParent = tree.children[0];
  const bgImageContent = imageParent.children;

  // The content overlay [0] and the color overlay [1]
  // should exist
  expect(bgImageContent.length).toBe(2);

  const colorOverlay = bgImageContent[1];
  const colorOverlayMainStyle = colorOverlay.props.style[0];
  expect(colorOverlayMainStyle.backgroundColor).toBe('red');
  // Needs to be 100%
  expect(colorOverlayMainStyle.width).toBe('100%');
  // Will inheriet on state change, so must not be defined
  expect(colorOverlayMainStyle.height).toBeUndefined;

  // Needs to be placed on top of content: z-1 + absolute
  expect(colorOverlay.props.style[1].zIndex).toBe(1)
  expect(colorOverlay.props.style[2].position).toBe('absolute')
});


test('PRIVATE FUNC: createPositionStyle should create a style', () => {
  expect(_createPositionStyle(2)).toEqual({
    top: 0,
    zIndex: 2
  });
});


test('SNAPSHOT: All functionality should match prev snapshot', () => {
  const tree = renderer.create(
       <Hero
     source={{uri: imageUrl}}
    renderOverlay={() => <Text>React Native Hero</Text>}
       colorOverlay={'red'}
    colorOpacity={.40}
    fullWidth={false}
    minHeight={1000} />
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
