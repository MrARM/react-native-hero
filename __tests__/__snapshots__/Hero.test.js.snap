// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SNAPSHOT: All functionality should match prev snapshot 1`] = `
<View
  style={
    Object {
      "opacity": 1,
    }
  }
>
  <Image
    resizeMode="cover"
    source={
      Object {
        "uri": "https://catsarethebest.com/cat1",
      }
    }
    style={
      Object {
        "height": 1000,
        "width": "100%",
      }
    }
  >
    <View
      onLayout={[Function]}
      style={
        Array [
          Object {
            "backgroundColor": "transparent",
          },
          Object {
            "top": 0,
            "zIndex": 2,
          },
        ]
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
      >
        React Native Hero
      </Text>
    </View>
    <View
      style={
        Array [
          Object {
            "backgroundColor": "red",
            "height": 1000,
            "opacity": 0.4,
            "width": "100%",
          },
          Object {
            "top": 0,
            "zIndex": 1,
          },
          Object {
            "position": "absolute",
          },
        ]
      }
    />
  </Image>
</View>
`;
